### PR TITLE
Refactor WKScriptMessage handling.

### PR DESCRIPTION
**Pull request by montehurd**
Broke things up into individual methods.

Added WMFWKScriptMessageType so we can more easily specify and check for expected body message types.

Fixed bug introduced with the recent null check fixes which prevented peekElement from being properly reset.

Fixes bug preventing reference panel from scrolling all the way to the bottom if there's a lot of text in the panel.

Fixes bug with reference links being peekable - they should show the panel instead.

### CODE REVIEW

File: Wikipedia/assets/index.js [hunk 1]
```diff
@@ -124,7 +124,7 @@ function maybeSendMessageForTarget(event, hrefTarget){
     var href = hrefTarget.getAttribute( "href" );
     var hrefClass = hrefTarget.getAttribute('class');
     if (hrefTarget.getAttribute( "data-action" ) === "edit_section") {
-        window.webkit.messageHandlers.clicks.postMessage({"editClicked": { sectionId: hrefTarget.getAttribute( "data-id" ) }});
```
Comment (montehurd) [inline Wikipedia/assets/index.js]: This PR makes the JS not send dictionaries within dictionaries. 
Crashes were happening when these nested dictionaries were NSNull and we tried to access values on them with dictionary subscript syntax - i.e. `dict[@"somekey"]`

File: Wikipedia/Code/WebViewController.m [hunk 2]
```diff
@@ -89,146 +90,164 @@ - (instancetype)initWithSession:(SessionSingleton*)aSession {
...
-     *  All messages from JS can contain NSNulls.
-     *  All elements must be checked for existence of NSNull before being passed on to any other methods.
-     * Additionally body is untyped, so its class must be
-     * checked before proceeding.
-     */
-
-    if ([message.body isKindOfClass:[NSNull class]]) {
+
+    WMFWKScriptMessageType messageType = [WKScriptMessage wmf_typeForMessageName:message.name];
+    if (![message.body isKindOfClass:[WKScriptMessage wmf_expectedMessageBodyClassForType:messageType]]) {
```
Comment (montehurd) [inline Wikipedia/Code/WebViewController.m]: All handler methods will be passed correct type - never NSNull.

**Comment by montehurd** [comment]
@joewalsh @coreyfloyd should be good for review.

File: Wikipedia/Code/WebViewController.m [hunk 3]
```diff
@@ -89,146 +90,164 @@ - (instancetype)initWithSession:(SessionSingleton*)aSession {
...
+            [self handleArticleStateScriptMessage:message.body];
+            break;
+        case WMFWKScriptMessageUnknown:
+            NSAssert(NO, @"Unhandled script message type!");
+            break;
+    }
+}
+
+- (void)handlePeekScriptMessage:(NSDictionary*)messageDict {
+    if(messageDict.count == 0){
```
Comment (joewalsh) [inline Wikipedia/Code/WebViewController.m]: Minor suggestion - check for the elements you want (tagName, src, href) rather than checking messageCount == 0 - this way the behavior is safe even if some objects are missing or the wrong keys are passed

**Comment by joewalsh** [comment]
👍

File: Wikipedia/Code/WebViewController.m (hunk 3 above)
Comment (montehurd) [inline Wikipedia/Code/WebViewController.m]: Ah, good point.

File: Wikipedia/Code/WKScriptMessage+WMFScriptMessage.m [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        case WMFWKScriptMessageLateJavascriptTransform:
+            return [NSString class];
+            break;
+        case WMFWKScriptMessageArticleState:
+            return [NSString class];
+            break;
+        case WMFWKScriptMessageConsoleMessage:
+            return [NSDictionary class];
+            break;
+        case WMFWKScriptMessageUnknown:
```
Comment (coreyfloyd) [inline Wikipedia/Code/WKScriptMessage+WMFScriptMessage.m]: This is better for the "0" case - because that way its the default which makes it safer

File: Wikipedia/Code/WMFRotationRespectingTabBarController.swift [hunk 5]
```diff
@@ -1,6 +1,18 @@
 
 import UIKit
 
+//HAX: work-around for iOS 9 infinite recursion loop bug with UIAlertController
+// http://stackoverflow.com/a/36476346/135557 . Noticed while attempting 3d touch
+// on link in reference panel.
+extension UIAlertController {
```
Comment (coreyfloyd) [inline Wikipedia/Code/WMFRotationRespectingTabBarController.swift]: This needs to be in a separate PR

File: Wikipedia/Code/WebViewController.m (hunk 2 above)
Comment (coreyfloyd) [inline Wikipedia/Code/WebViewController.m]: Ask don't tell

**Comment by joewalsh** [comment]
👍 after the UIAlertController stuff is in a separate PR

**Comment by montehurd** [comment]
I moved the "ref panel scroll issue" bits to https://github.com/wikimedia/wikipedia-ios/pull/773

**Comment by montehurd** [comment]
I moved the "ref panel link 3d touch crash fix" to #774

**Comment by montehurd** [comment]
@coreyfloyd I made those changes. Should be good to go.

**Comment by montehurd** [comment]
Note: the UIAlertController bits ended up being redundant. https://github.com/wikimedia/wikipedia-ios/pull/774 fixed same issue.
